src/actions/EditNew.ts

```diff
@@ -104,70 +77,112 @@ class EditNew implements Action {
     return { thatMark };
   }
 
-  async runDelimiterTargets(
-    editor: TextEditor,
-    commandTargets: CommandTarget[],
-    delimiterTargets: DelimiterTarget[]
-  ) {
+  async runCommandTargets(editor: TextEditor, state: State): Promise<State> {
+    const commandTargets: CommandTarget[] = state.targets
+      .map((target, index) => {
+        const context = target.getEditNewContext(this.isBefore);
+        if (context.type === "command") {
+          return {
+            target,
+            index,
+            command: context.command,
+          };
+        }
+      })
+      .filter((target): target is CommandTarget => !!target);
+
+    if (commandTargets.length === 0) {
+      return state;
+    }
+
+    const command = ensureSingleCommand(commandTargets);
+
+    await this.setSelections(commandTargets.map(({ target }) => target));
+
+    const [updatedTargetRanges, updatedThatRanges] =
+      await callFunctionAndUpdateRanges(
+        this.graph.rangeUpdater,
+        () => commands.executeCommand(command),
+        editor.document,
+        [
+          state.targets.map(({ contentRange }) => contentRange),
+          state.thatRanges,
+        ]
+      );
+
+    const cursorRanges = [...state.cursorRanges];
+    commandTargets.forEach((commandTarget, index) => {
+      cursorRanges[commandTarget.index] = editor.selections[index];
+    });
+
+    return {
+      targets: state.targets.map((target, index) =>
+        target.withContentRange(updatedTargetRanges[index])
+      ),
+      thatRanges: updatedThatRanges,
+      cursorRanges,
+    };
+  }
+
+  async runDelimiterTargets(editor: TextEditor, state: State): Promise<State> {
+    const delimiterTargets: DelimiterTarget[] = state.targets
+      .map((target, index) => {
+        const context = target.getEditNewContext(this.isBefore);
+        if (context.type === "delimiter") {
+          return {
+            target,
+            index,
+          };
+        }
+      })
+      .filter((target): target is DelimiterTarget => !!target);
+
+    if (delimiterTargets.length === 0) {
+      return state;
+    }
+
     const position = this.isBefore ? "before" : "after";
-    const edits = delimiterTargets.flatMap((target) =>
+    const edits = delimiterTargets.map((target) =>
       toPositionTarget(target.target, position).constructChangeEdit("")
     );
 
-    const cursorSelections = { selections: editor.selections };
-    const contentSelections = {
-      selections: delimiterTargets.map(
-        ({ target }) => target.thatTarget.contentSelection
-      ),
+    const thatSelections = {
+      selections: state.thatRanges.map(toSelection),
+    };
+
+    const cursorSelections = {
+      selections: state.cursorRanges.map(toSelection),
```

Comment:
So the only real selections here are command targets; the ranges corresponding to delimiter targets will all be dummy `new Range(0, 0, 0, 0)` ranges?  Worth a comment

I'm also not in love with the dummy ranges; it's a bit confusing, and it's wasted computation (not that that need be a primary concern here)

I would be inclined to have the cursor ranges in state be `undefined` until we have a real value for them

Here, you could construct a list which only includes existing cursor selections that are not `undefined`, and keep indices back into the original state for after the range update